src/ripple/app/tx/impl/Transactor.cpp

```diff
@@ -673,4 +618,173 @@ TER Transactor::checkMultiSign ()
+                "Transaction serdes mismatch";
+            JLOG(j_.info) << to_string(mTxn.getJson (0));
+            JLOG(j_.fatal) << s2.getJson (0);
+            assert (false);
+        }
+    }
+#endif
+
+    TER terResult = apply();
+
+    if (terResult == temUNKNOWN)
+    {
+        JLOG(j_.warning) <<
+            "applyTransaction: Invalid transaction: unknown transaction type";
+        auto const result =
+            std::make_pair(temUNKNOWN, false);
+        log(result, j_);
+        return result;
+    }
+
+    if (j_.debug)
+    {
+        std::string strToken;
+        std::string strHuman;
+
+        transResultInfo (terResult, strToken, strHuman);
+
+        j_.debug <<
+            "applyTransaction: terResult=" << strToken <<
+            " : " << terResult <<
+            " : " << strHuman;
+    }
+
+    bool didApply = isTesSuccess (terResult);
+
+    if (isTecClaim (terResult) && !(view().flags() & tapRETRY))
+    {
+        // only claim the transaction fee
+        JLOG(j_.debug) <<
+            "Reprocessing tx " << txID << " to only claim fee";
+        
+        ctx_.discard();
+
+        auto const txnAcct = view().peek(
+            keylet::account(mTxn.getAccountID(sfAccount)));
+
+        if (txnAcct)
+        {
+            std::uint32_t t_seq = mTxn.getSequence ();
+            std::uint32_t a_seq = txnAcct->getFieldU32 (sfSequence);
+
+            if (a_seq < t_seq)
+                terResult = terPRE_SEQ;
+            else if (a_seq > t_seq)
+                terResult = tefPAST_SEQ;
+            else
+            {
+                STAmount fee        = mTxn.getTransactionFee ();
+                STAmount balance    = txnAcct->getFieldAmount (sfBalance);
+
+                // We retry/reject the transaction if the account
+                // balance is zero or we're applying against an open
+                // ledger and the balance is less than the fee
+                if ((balance == zero) ||
+                    (view().open() && (balance < fee)))
+                {
+                    // Account has no funds or ledger is open
+                    terResult = terINSUF_FEE_B;
+                }
+                else
+                {
+                    if (fee > balance)
+                        fee = balance;
+                    txnAcct->setFieldAmount (sfBalance, balance - fee);
+                    txnAcct->setFieldU32 (sfSequence, t_seq + 1);
+                    view().update (txnAcct);
+                    didApply = true;
+                }
+            }
+        }
+        else
+        {
+            terResult = terNO_ACCOUNT;
+        }
+    }
+    else if (!didApply)
+    {
+        JLOG(j_.debug) << "Not applying transaction " << txID;
+    }
+
+    if (didApply)
+    {
+        // Transaction succeeded fully or (retries are
+        // not allowed and the transaction could claim a fee)
+
+        if(view().closed())
+        {
+            // VFALCO Fix this nonsense with Amount
```

Comment:
I'm unclear what you mean by "charge whatever fee they specified." We are guaranteed that `Amount` is XRP and charge them whatever the fee is.

I agree that we're using special knowledge: the `Amount` class (decoupled from `STAmount`) will come in handy here.